text/3778-inline-intents.md

```diff
@@ -0,0 +1,189 @@
+- Feature Name: `inline_intents`
+- Start Date: 2025-02-22
+- RFC PR: [rust-lang/rfcs#3778](https://github.com/rust-lang/rfcs/pull/3778)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add `#[inline(trampoline)]` and `#[inline(rarely)]` to give additional control
+over the inlining decisions made by the compiler.
+
+# Motivation
+[motivation]: #motivation
+
+Right now it's pretty common to just slap `#[inline]` on things without thinking
+about it all that hard about it, and the existing controls aren't great.
+
+Often we'll get PRs using `inline(always)` "because it's just calling something
+else so of course it should be inlined", for example.  But because of the
+bottom-up nature of inlining, that's a bad thing to do because if the callee
+happens to get inlined, then it'll "always" inline that callee too, which might
+not be what was actually desired.
```

Comment:
This paragraph is unclear to me; in particular, it refers to “the callee” and “that callee” and I am not sure which two (or one?) functions are being referred to. I am familiar with the idea that LLVM inlining proceeds as, if A calls B calls C, "inline C into B, and only then decide if B’s code would be good to inline into A", and I think that’s the general scenario this is referring to, but the paragraph should be comprehensible to people who haven’t met that model, and it’s not clear which of these functions we are supposing is `#[inline(always)]`.